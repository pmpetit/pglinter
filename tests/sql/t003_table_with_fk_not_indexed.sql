-- Example to demonstrate missing index on foreign key and T003 rule detection
CREATE EXTENSION pglinter;

\pset pager off

-- Create tables: parent with PK, child with FK but no index on FK
CREATE TABLE parent (
    id INTEGER PRIMARY KEY,
    name TEXT
);

CREATE TABLE child (
    child_id INTEGER PRIMARY KEY,
    parent_id INTEGER NOT NULL,
    value TEXT,
    CONSTRAINT fk_parent FOREIGN KEY (parent_id) REFERENCES parent(id)
);

-- No index on child.parent_id (foreign key column)

-- Insert some test data
INSERT INTO parent (id, name) VALUES (1, 'A'), (2, 'B');
INSERT INTO child (child_id, parent_id, value) VALUES (10, 1, 'foo'), (11, 2, 'bar');

-- Update table statistics
ANALYZE parent;
ANALYZE child;



-- Disable all rules first to isolate T004 testing
SELECT pglinter.disable_all_rules() AS all_rules_disabled;

-- Enable T003 specifically
SELECT pglinter.enable_rule('T003') AS t003_enabled;

-- Run table check to detect FKs without index
SELECT pglinter.perform_table_check(); -- Should include T004 results

-- Now let's fix the missing index issue
CREATE INDEX idx_child_parent_id ON child(parent_id);

-- Update statistics after schema changes
ANALYZE child;

-- Run T004 check again (should show no issues now)
SELECT pglinter.perform_table_check();


DROP TABLE child CASCADE;
DROP TABLE parent CASCADE;

DROP EXTENSION pglinter CASCADE;
