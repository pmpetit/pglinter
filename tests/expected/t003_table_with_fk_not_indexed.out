-- Example to demonstrate missing index on foreign key and T003 rule detection
CREATE EXTENSION pglinter;
\pset pager off
-- Create tables: parent with PK, child with FK but no index on FK
CREATE TABLE parent (
    id INTEGER PRIMARY KEY,
    name TEXT
);
CREATE TABLE child (
    child_id INTEGER PRIMARY KEY,
    parent_id INTEGER NOT NULL,
    value TEXT,
    CONSTRAINT fk_parent FOREIGN KEY (parent_id) REFERENCES parent(id)
);
-- No index on child.parent_id (foreign key column)
-- Insert some test data
INSERT INTO parent (id, name) VALUES (1, 'A'), (2, 'B');
INSERT INTO child (child_id, parent_id, value) VALUES (10, 1, 'foo'), (11, 2, 'bar');
-- Update table statistics
ANALYZE parent;
ANALYZE child;
-- Disable all rules first to isolate T004 testing
SELECT pglinter.disable_all_rules() AS all_rules_disabled;
NOTICE:  üî¥ Disabled 21 rule(s)
 all_rules_disabled 
--------------------
                 21
(1 row)

-- Enable T003 specifically
SELECT pglinter.enable_rule('T003') AS t003_enabled;
NOTICE:  ‚úÖ Rule T003 has been enabled
 t003_enabled 
--------------
 t
(1 row)

-- Run table check to detect FKs without index
SELECT pglinter.perform_table_check(); -- Should include T004 results
NOTICE:  üîç pglinter found 1 issue(s):
NOTICE:  ==================================================
NOTICE:  ‚ö†Ô∏è  [T003] WARNING: TABLE unindexed constraint 1 : 
public.child constraint name:fk_parent 

NOTICE:  ==================================================
NOTICE:  üìä Summary: 0 error(s), 1 warning(s), 0 info
NOTICE:  üü° Some warnings found - consider reviewing for optimization
 perform_table_check 
---------------------
 t
(1 row)

-- Now let's fix the missing index issue
CREATE INDEX idx_child_parent_id ON child(parent_id);
-- Update statistics after schema changes
ANALYZE child;
-- Run T004 check again (should show no issues now)
SELECT pglinter.perform_table_check();
NOTICE:  ‚úÖ No issues found - database schema looks good!
 perform_table_check 
---------------------
 t
(1 row)

DROP TABLE child CASCADE;
DROP TABLE parent CASCADE;
DROP EXTENSION pglinter CASCADE;
